File: app/src/main/java/activity/ListaGenericaActivity.java
package activity;

import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ProgressBar;

import java.util.HashMap;
import java.util.Map;
import java.util.Random;

import br.com.icaro.filme.R;
import domain.ListaJava;
import utils.Constantes;
import utils.UtilsApp;

/**
 * Created by icaro on 04/10/16.
 */
public class ListaGenericaActivity  extends BaseActivity{

    private RecyclerView recyclerView;
    private ProgressBar progressBar;
    private String list_id;
    private String old = "";
    private Map<String, String> map = new HashMap<>();
    private ListaJava listaJava = null;
    private SwipeRefreshLayout swipeRefreshLayout;
    private String TAG = this.getClass().getName();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_lista);
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        setUpToolBar();
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setTitle(getIntent().getStringExtra(Constantes.INSTANCE.getLISTA_GENERICA()));
        list_id = getIntent().getStringExtra(Constantes.INSTANCE.getLISTA_ID());
        map = (Map<String, String>) getIntent().getSerializableExtra(Constantes.INSTANCE.getBUNDLE());
        progressBar = (ProgressBar) findViewById(R.id.progress);
        swipeRefreshLayout = (SwipeRefreshLayout) findViewById(R.id.swipeToRefresh);
        recyclerView = (RecyclerView) findViewById(R.id.recycleView_favorite);
        recyclerView.setLayoutManager(new GridLayoutManager(this, 3));
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.setHasFixedSize(true);

        if (map == null){
            swipeRefreshLayout.setEnabled(false);
        }
        swipeRefreshLayout.setColorSchemeResources(R.color.primary, R.color.primary_dark, R.color.accent);
        swipeRefreshLayout.setOnRefreshListener(OnRefreshListener());

//        AdView adview = (AdView) findViewById(R.id.adView);
//        AdRequest adRequest = new AdRequest.Builder()
//                .build();
//        adview.loadAd(adRequest);

    }

    protected SwipeRefreshLayout.OnRefreshListener OnRefreshListener() {
        return new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                String numero = String.valueOf(new Random().nextInt(10));
                list_id = map.get("id"+numero);
                if (old.equals(list_id)){
                    if (list_id.equals("11060")) {
                        list_id = "10"; //Top 50 Grossing Films of All Time (Worldwide)
                    } else {
                        list_id = "11060"; // PopCorn Favorites
                    }
                }

                if(UtilsApp.isNetWorkAvailable(ListaGenericaActivity.this)) {
                 //   new TMDVAsync().execute();
                }
            }
        };
    }

    @Override
    protected void onResume() {
        super.onResume();
        if(UtilsApp.isNetWorkAvailable(this)) {
        //    new TMDVAsync().execute();
        }
    }
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        return true;
    }

//    private class TMDVAsync extends AsyncTask<Void, Void, Void> {
//
//        @Override
//        protected void onPreExecute() {
//            super.onPreExecute();
//           // Log.d(TAG, list_id);
//        }
//
//        @Override
//        protected Void doInBackground(Void... voids) {
//            //if (listaJava == null) {
//                try {
//                    listaJava = null;
//                    listaJava = FilmeService.getLista(list_id);
//                    //Metodos criados. Tudo gambiara. Precisa arrumar - 11060
//                    if (listaJava.getItems() != null) {
//                        Collections.sort(listaJava.getItems());
//                    }
//                } catch (Exception e){
//                    FirebaseCrash.report(e);
//                    runOnUiThread(new Runnable() {
//                        @Override
//                        public void run() {
//                            Toast.makeText(ListaGenericaActivity.this, R.string.ops, Toast.LENGTH_SHORT).show();
//                        }
//                    });
//                }
//            //}
//
//            return null;
//        }
//
//        @Override
//        protected void onPostExecute(Void aVoid) {
//            super.onPostExecute(aVoid);
//                if (listaJava != null) {
//                    swipeRefreshLayout.setRefreshing(false);
//                    getSupportActionBar().setTitle(listaJava.getName());
//                    progressBar.setVisibility(View.GONE);
//                    recyclerView.setAdapter(new ListUserAdapter(ListaGenericaActivity.this,
//                            listaJava != null ? listaJava.items : null));
//                    old = list_id;
//                }
//        }
//    }
}

File: app/src/main/java/activity/ListaGenericaActivity.kt
package activity

import adapter.ListUserAdapter
import android.content.pm.ActivityInfo
import android.os.Bundle
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.GridLayoutManager
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import br.com.icaro.filme.R
import domain.API
import kotlinx.android.synthetic.main.activity_lista.*
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subscriptions.CompositeSubscription
import utils.Constantes
import utils.InfiniteScrollListener
import utils.UtilsApp
import java.util.*

/**
 * Created by icaro on 04/10/16.
 */
class ListaGenericaActivity : BaseActivity() {


    private lateinit var list_id: String
    private var totalPagina: Int = 1
    private var pagina = 1
    private lateinit var map: Map<String, String>
    private val TAG = this.javaClass.name


    private var subscriptions = CompositeSubscription()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_lista)
        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        setUpToolBar()
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.title = intent.getStringExtra(Constantes.LISTA_GENERICA)
        list_id = intent.getStringExtra(Constantes.LISTA_ID)
        if (intent.hasExtra(Constantes.BUNDLE)) {
            map =  HashMap()
            map = intent.getSerializableExtra(Constantes.BUNDLE) as Map<String, String>
        }
        recycleView_favorite.apply {
            val gridlayout = GridLayoutManager(this@ListaGenericaActivity, 3);
            layoutManager = gridlayout
            itemAnimator = DefaultItemAnimator()
            setHasFixedSize(true)
            addOnScrollListener(InfiniteScrollListener({ getLista() }, gridlayout))
            adapter = ListUserAdapter(this@ListaGenericaActivity)
        }

        //        AdView adview = (AdView) findViewById(R.id.adView);
        //        AdRequest adRequest = new AdRequest.Builder()
        //                .build();
        //        adview.loadAd(adRequest);

    }

    override fun onResume() {
        super.onResume()
        subscriptions = CompositeSubscription()
        if (UtilsApp.isNetWorkAvailable(this)) {
            getLista()
        }
    }

    override fun onPause() {
        super.onPause()
        subscriptions.clear()
    }


    fun getLista() {
        if (totalPagina >= pagina) {
            val inscricao = API(context = this).getLista(id = list_id, pagina = pagina)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({
                        (recycleView_favorite.adapter as ListUserAdapter).addItens(it.results, it?.totalResults!!)
                        pagina = it.page!!
                        totalPagina = it.totalPages!!
                        ++pagina
                    }, { erro ->
                        Toast.makeText(this, getString(R.string.ops), Toast.LENGTH_LONG).show()
                        Log.d(javaClass.simpleName, "Erro " + erro.message)
                    })
            subscriptions.add(inscricao)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {

        when (item.itemId) {
            android.R.id.home -> finish()
            R.id.nova_lista -> {
                val numero = Random().nextInt(10).toString()
                supportActionBar?.title = map?.get("title$numero")
                list_id = map?.get("id$numero").toString()
                recycleView_favorite.adapter = ListUserAdapter(this)
                pagina = 1
                totalPagina = 1
                getLista()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        if (intent.hasExtra(Constantes.BUNDLE) )
        menuInflater.inflate(R.menu.menu_random_lista, menu)

        return true
    }

}
